File: app/src/main/java/com/example/map/data/remote/api/KakaoLocalServiceFactory.java
package com.example.map.data.remote.api;

import androidx.annotation.NonNull;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class KakaoLocalServiceFactory {
    private static final String BASE_URL = "https://dapi.kakao.com";
    private static final String REST_API_KEY = "b83c1687f3a7d168dbd29d37e543d9c1";

    public static <T> T create(Class<T> apiClass) {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient client = new OkHttpClient.Builder()
            .addInterceptor(new HeaderInterceptor())
            .addInterceptor(loggingInterceptor)
            .build();
        return new Retrofit.Builder()
            .baseUrl(BASE_URL)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(apiClass);
    }

    private static class HeaderInterceptor implements Interceptor {

        @NonNull
        @Override
        public Response intercept(@NonNull Chain chain) throws IOException {
            Request request = chain.request().newBuilder()
                .addHeader("Authorization", "KakaoAK " + REST_API_KEY)
                .build();
            return chain.proceed(request);
        }
    }
}

File: app/src/main/java/com/example/map/data/remote/api/KakaoLocalServiceFactory.kt
package com.example.map.data.remote.api

import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException

object KakaoLocalServiceFactory {
    private const val BASE_URL = "https://dapi.kakao.com"
    private const val REST_API_KEY = "b83c1687f3a7d168dbd29d37e543d9c1"
    fun <T> create(apiClass: Class<T>): T {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        val client: OkHttpClient = OkHttpClient.Builder()
            .addInterceptor(HeaderInterceptor())
            .addInterceptor(loggingInterceptor)
            .build()
        return Retrofit.Builder()
            .baseUrl(BASE_URL)
            .client(client)
            .addConverterFactory(
                GsonConverterFactory.create(
                    GsonBuilder().setFieldNamingPolicy(
                        FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES
                    ).create()
                )
            ).build()
            .create(apiClass)
    }

    private class HeaderInterceptor : Interceptor {
        @Throws(IOException::class)
        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request().newBuilder()
                .addHeader("Authorization", "KakaoAK $REST_API_KEY")
                .build()
            return chain.proceed(request)
        }
    }
}
